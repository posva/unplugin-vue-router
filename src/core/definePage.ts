import {
  generateTransform,
  isCallOf,
  parseSFC,
  MagicString,
  checkInvalidScopeReference,
} from '@vue-macros/common'
import { Thenable, TransformResult } from 'unplugin'
import type {
  CallExpression,
  Node,
  ObjectProperty,
  Statement,
  StringLiteral,
} from '@babel/types'
import { walkAST } from 'ast-walker-scope'
import { CustomRouteBlock } from './customBlock'
import { warn } from './utils'

const MACRO_DEFINE_PAGE = 'definePage'
const MACRO_DEFINE_PAGE_QUERY = /[?&]definePage\b/

function isStringLiteral(node: Node | null | undefined): node is StringLiteral {
  return node?.type === 'StringLiteral'
}

export function definePageTransform({
  code,
  id,
}: {
  code: string
  id: string
}): Thenable<TransformResult> {
  if (!code.includes(MACRO_DEFINE_PAGE)) return

  const sfc = parseSFC(code, id)
  if (!sfc.scriptSetup) return

  // are we extracting only the definePage object
  const isExtractingDefinePage = MACRO_DEFINE_PAGE_QUERY.test(id)

  const { script, scriptSetup, getSetupAst } = sfc
  const setupAst = getSetupAst()

  const definePageNodes = (setupAst?.body || ([] as Node[]))
    .map((node) => {
      if (node.type === 'ExpressionStatement') node = node.expression
      return isCallOf(node, MACRO_DEFINE_PAGE) ? node : null
    })
    .filter((node): node is CallExpression => !!node)

  if (!definePageNodes.length) {
    return isExtractingDefinePage
      ? // e.g. index.vue?definePage that contains a commented `definePage()
        'export default {}'
      : // e.g. index.vue that contains a commented `definePage()
        null
  } else if (definePageNodes.length > 1) {
    throw new SyntaxError(`duplicate definePage() call`)
  }

  const definePageNode = definePageNodes[0]
  const setupOffset = scriptSetup.loc.start.offset

  // we only want the page info
  if (isExtractingDefinePage) {
    const s = new MagicString(code)
    // remove everything except the page info

    const routeRecord = definePageNode.arguments[0]

    const scriptBindings = setupAst?.body ? getIdentifiers(setupAst.body) : []

    checkInvalidScopeReference(routeRecord, MACRO_DEFINE_PAGE, scriptBindings)

    // NOTE: this doesn't seem to be any faster than using MagicString
    // return (
    //   'export default ' +
    //   code.slice(
    //     setupOffset + routeRecord.start!,
    //     setupOffset + routeRecord.end!
    //   )
    // )

    s.remove(setupOffset + routeRecord.end!, code.length)
    s.remove(0, setupOffset + routeRecord.start!)
    s.prepend(`export default `)

    return generateTransform(s, id)
  } else {
    // console.log('!!!', definePageNode)

    const s = new MagicString(code)

    // s.removeNode(definePageNode, { offset: setupOffset })
    s.remove(
      setupOffset + definePageNode.start!,
      setupOffset + definePageNode.end!
    )

    return generateTransform(s, id)
  }
}

export function extractDefinePageNameAndPath(
  sfcCode: string,
  id: string
): { name?: string; path?: string } | null | undefined {
  if (!sfcCode.includes(MACRO_DEFINE_PAGE)) return

  const sfc = parseSFC(sfcCode, id)

  if (!sfc.scriptSetup) return

  const { getSetupAst } = sfc
  const setupAst = getSetupAst()

  const definePageNodes = (setupAst?.body ?? ([] as Node[]))
    .map((node) => {
      if (node.type === 'ExpressionStatement') node = node.expression
      return isCallOf(node, MACRO_DEFINE_PAGE) ? node : null
    })
    .filter((node): node is CallExpression => !!node)

  if (!definePageNodes.length) {
    return
  } else if (definePageNodes.length > 1) {
    throw new SyntaxError(`duplicate definePage() call`)
  }

  const definePageNode = definePageNodes[0]

  const routeRecord = definePageNode.arguments[0]
  if (routeRecord.type !== 'ObjectExpression') {
    throw new SyntaxError(
      `[${id}]: definePage() expects an object expression as its only argument`
    )
  }

  const routeInfo: Pick<CustomRouteBlock, 'name' | 'path'> = {}

  for (const prop of routeRecord.properties) {
    if (prop.type === 'ObjectProperty' && prop.key.type === 'Identifier') {
      if (prop.key.name === 'name') {
        if (prop.value.type !== 'StringLiteral') {
          warn(`route name must be a string literal. Found in "${id}".`)
        } else {
          routeInfo.name = prop.value.value
        }
      } else if (prop.key.name === 'path') {
        if (prop.value.type !== 'StringLiteral') {
          warn(`route path must be a string literal. Found in "${id}".`)
        } else {
          routeInfo.path = prop.value.value
        }
      }
    }
  }

  return routeInfo
}

function extractRouteAlias(
  aliasValue: ObjectProperty['value'],
  id: string
): string[] | undefined {
  if (
    aliasValue.type !== 'StringLiteral' &&
    aliasValue.type !== 'ArrayExpression'
  ) {
    warn(`route alias must be a string literal. Found in "${id}".`)
  } else {
    return aliasValue.type === 'StringLiteral'
      ? [aliasValue.value]
      : aliasValue.elements.filter(isStringLiteral).map((el) => el.value)
  }
}

const getIdentifiers = (stmts: Statement[]) => {
  let ids: string[] = []
  walkAST(
    {
      type: 'Program',
      body: stmts,
      directives: [],
      sourceType: 'module',
      sourceFile: '',
    },
    {
      enter(node) {
        if (node.type === 'BlockStatement') {
          this.skip()
        }
      },
      leave(node) {
        if (node.type !== 'Program') return
        ids = Object.keys(this.scope)
      },
    }
  )

  return ids
}
