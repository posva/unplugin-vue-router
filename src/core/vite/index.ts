import { type ViteDevServer } from 'vite'
import { type ServerContext } from '../../options'
import { MODULE_ROUTES_PATH, asVirtualId } from '../moduleConstants'

export function createViteContext(server: ViteDevServer): ServerContext {
  function invalidate(path: string) {
    const { moduleGraph } = server
    const foundModule = moduleGraph.getModuleById(path)
    if (foundModule) {
      moduleGraph.invalidateModule(foundModule, undefined, undefined, true)
      // for (const mod of foundModule.importers) {
      //   console.log(`Invalidating ${mod.url}`)
      //   moduleGraph.invalidateModule(mod)
      // }
      setTimeout(() => {
        console.log(`Sending update for ${foundModule.url}`)
        server.ws.send({
          type: 'update',
          updates: [
            {
              acceptedPath: path,
              path: path,
              // NOTE: this was in the
              // timestamp: ROUTES_LAST_LOAD_TIME.value,
              timestamp: Date.now(),
              type: 'js-update',
            },
          ],
        })
      }, 100)
    }
    return !!foundModule
  }

  function reload() {
    server.ws.send({
      type: 'full-reload',
      path: '*',
    })
  }

  /**
   * Triggers HMR for the vue-router/auto-routes module.
   */
  async function updateRoutes() {
    const modId = asVirtualId(MODULE_ROUTES_PATH)
    const mod = server.moduleGraph.getModuleById(modId)
    if (mod) {
      return server.reloadModule(mod)
    }
  }

  return {
    invalidate,
    updateRoutes,
    reload,
  }
}
