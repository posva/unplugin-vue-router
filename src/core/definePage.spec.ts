import { TransformResult } from 'vite'
import { expect, describe, it } from 'vitest'
import { definePageTransform, extractDefinePageNameAndPath } from './definePage'

const sampleCode = `
<script setup>
const a = 1
definePage({
  name: 'custom',
  path: '/custom',
})
const b = 1
</script>

<template>
  <div>hello</div>
</template>
      `

describe('definePage', () => {
  it('removes definePage', async () => {
    const result = (await definePageTransform({
      code: sampleCode,
      id: 'src/pages/basic.vue',
    })) as Exclude<TransformResult, string>

    expect(result).toHaveProperty('code')
    expect(result?.code).toMatchSnapshot()
  })

  describe('imports', () => {
    it('keeps used named imports', async () => {
      const result = (await definePageTransform({
        code: `
<script setup>
import { my_var, not_used, my_func, my_num } from './lib'
definePage({
  meta: {
    [my_var]: 'hello',
    other: my_func,
    custom() {
      return my_num
    }
  }
})
</script>
`,
        id: 'src/pages/with-imports.vue&definePage&vue&lang.ts',
      })) as Exclude<TransformResult, string>
      expect(result).toHaveProperty('code')
      expect(result?.code).toMatchSnapshot()
    })

    it('keeps used default imports', async () => {
      const result = (await definePageTransform({
        code: `
<script setup>
import my_var from './lib'
definePage({
  meta: {
    [my_var]: 'hello',
  }
  })
</script>
`,
        id: 'src/pages/with-imports.vue&definePage&vue&lang.ts',
      })) as Exclude<TransformResult, string>
      expect(result).toHaveProperty('code')
      expect(result?.code).toMatchSnapshot()
    })

    it('removes default import if not used', async () => {
      const result = (await definePageTransform({
        code: `
<script setup>
import my_var from './lib'
definePage({name: 'ok'})
</script>
`,
        id: 'src/pages/with-imports.vue&definePage&vue&lang.ts',
      })) as Exclude<TransformResult, string>
      expect(result).toHaveProperty('code')
      expect(result?.code).toMatchSnapshot()
    })

    it('works with star imports', async () => {
      const result = (await definePageTransform({
        code: `
<script setup>
import * as lib from './my-lib'
definePage({
  meta: {
    [lib.my_var]: 'hello',
  }
  })
</script>
`,
        id: 'src/pages/with-imports.vue&definePage&vue&lang.ts',
      })) as Exclude<TransformResult, string>
      expect(result).toHaveProperty('code')
      expect(result?.code).toMatchSnapshot()
    })

    it('removes star imports if not used', async () => {
      const result = (await definePageTransform({
        code: `
<script setup>
import * as lib from './my-lib'
definePage({name: 'ok'})
</script>
`,
        id: 'src/pages/with-imports.vue&definePage&vue&lang.ts',
      })) as Exclude<TransformResult, string>
      expect(result).toHaveProperty('code')
      expect(result?.code).toMatchSnapshot()
    })

    it('works when combining named and default imports', async () => {
      const result = (await definePageTransform({
        code: `
<script setup>
import my_var, { not_used, my_func, not_used_either } from './lib'
definePage({
  meta: {
    [my_var]: 'hello',
    other: my_func,
  }
})
</script>
`,
        id: 'src/pages/with-imports.vue&definePage&vue&lang.ts',
      })) as Exclude<TransformResult, string>
      expect(result).toHaveProperty('code')
      expect(result?.code).toMatchSnapshot()
    })
  })

  it('works with jsx', async () => {
    const code = `
    const a = 1
    definePage({
      name: 'custom',
      path: '/custom',
    })
    const b = 1
    `,
      result = (await definePageTransform({
        code,
        id: 'src/pages/basic.jsx?definePage&lang.jsx',
      })) as Exclude<TransformResult, string>
    expect(result).toBeDefined()
    expect(result).toHaveProperty('code')
    expect(result?.code).toMatchSnapshot()
  })

  it('throws if definePage uses a variable from the setup', async () => {
    const code = `
<script setup>
const a = 1
definePage({
  name: a,
})
</script>
`
    // the function syntax works with sync and async errors
    await expect(async () => {
      await definePageTransform({
        code,
        id: 'src/pages/basic.vue&definePage&vue',
      })
    }).rejects.toThrowError()
  })

  it('extracts name and path', async () => {
    expect(
      await extractDefinePageNameAndPath(sampleCode, 'src/pages/basic.vue')
    ).toEqual({
      name: 'custom',
      path: '/custom',
    })
  })

  it('extract name skipped when non existent', async () => {
    expect(
      await extractDefinePageNameAndPath(
        `
<script setup>
const a = 1
const b = 1
</script>

<template>
  <div>hello</div>
</template>
      `,
        'src/pages/basic.vue'
      )
    ).toBeFalsy()
  })

  it('works with comments', async () => {
    const code = `
<script setup>
// definePage
</script>

<template>
  <div>hello</div>
</template>
      `
    // no need to transform
    let result = (await definePageTransform({
      code,
      id: 'src/pages/basic.vue',
    })) as Exclude<TransformResult, string>
    expect(result).toBeFalsy()

    // should give an empty object
    result = (await definePageTransform({
      code,
      id: 'src/pages/basic.vue?definePage&vue',
    })) as Exclude<TransformResult, string>

    expect(result).toBe('export default {}')
  })

  it('works if file is named definePage', async () => {
    const result = (await definePageTransform({
      code: sampleCode,
      id: 'src/pages/definePage.vue',
    })) as Exclude<TransformResult, string>

    expect(result).toHaveProperty('code')
    // should be the sfc without the definePage call
    expect(result?.code).toMatchSnapshot()

    expect(
      await definePageTransform({
        code: sampleCode,
        id: 'src/pages/definePage.vue?definePage&vue',
      })
    ).toMatchObject({
      code: `\
export default {
  name: 'custom',
  path: '/custom',
}`,
    })

    expect(
      await extractDefinePageNameAndPath(sampleCode, 'src/pages/definePage.vue')
    ).toEqual({
      name: 'custom',
      path: '/custom',
    })
  })
})
