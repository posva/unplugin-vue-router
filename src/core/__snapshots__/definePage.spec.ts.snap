// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`definePage > imports > keeps used default imports 1`] = `
"import my_var from './lib'
export default {
  meta: {
    [my_var]: 'hello',
  }
  }"
`;

exports[`definePage > imports > keeps used named imports 1`] = `
"import {my_var, my_func, my_num} from './lib'
export default {
  meta: {
    [my_var]: 'hello',
    other: my_func,
    custom() {
      return my_num
    }
  }
}"
`;

exports[`definePage > imports > removes default import if not used 1`] = `"export default {name: 'ok'}"`;

exports[`definePage > imports > removes star imports if not used 1`] = `"export default {name: 'ok'}"`;

exports[`definePage > imports > works when combining named and default imports 1`] = `
"import my_var, {my_func} from './lib'
export default {
  meta: {
    [my_var]: 'hello',
    other: my_func,
  }
}"
`;

exports[`definePage > imports > works with star imports 1`] = `
"import * as lib from './my-lib'
export default {
  meta: {
    [lib.my_var]: 'hello',
  }
  }"
`;

exports[`definePage > removes definePage 1`] = `
"
<script setup>
const a = 1

const b = 1
</script>

<template>
  <div>hello</div>
</template>
      "
`;

exports[`definePage > works if file is named definePage 1`] = `
"
<script setup>
const a = 1

const b = 1
</script>

<template>
  <div>hello</div>
</template>
      "
`;

exports[`definePage > works with jsx 1`] = `
"export default {
      name: 'custom',
      path: '/custom',
    }"
`;
